### PR TITLE
Piranha Plant Rework (Redux): Adaptive Roots

### PR DESCRIPTION
**Pull request by H3ra-57**
This rework aims to turn Piranha Plant into a stance character with a new mechanic, Adaptive Roots. The standard plant has the best movement, the putrid plant is the most defensive with attacks capable of poisoning the foe, and the prickly plant packs the biggest punch with heavy, yet slow attacks.

### Entering Stances
- Simply taunt to change stance. Up taunt changes to the standard plant, down taunt changes to Putrid Plant, and side taunt changes to Prickly Plant.
### General Stance Changes
- Plant moves slower when Putrid/Prickly
- Putrid's attacks deal less damage, and bites deal poison DoT
- Prickly's attacks are slowed down, but deal more damage, head attacks most of all
### Ptooie
- Putrid Ptooie deals poison DoT
- Prickly Ptooie is larger and deals more damage, but goes a shorter distance
### Poison Breath → Adaptive Maw
Poison Cloud has been changed into a completely new move, Adaptive Maw- a close-range spit that has different properties depending on Plant's current stance!
- Standard Plant spits a gust of wind that sends at a nasty angle
- Putrid Plant spits a multihit poison breath that poisons
- Prickly Plant bites instead of spitting, dealing high shield damage and piercing shields at full charge
### Piranhacopter
- Is now a single hit
- Grounded version no longer leaves the ground
- Cancel goes into special fall immediately
- Aerial version moves much slower
### Changelog and hitbox diffs under construction
<<<=== PIRANHA PLANT===>>>
STATS/ADAPTIVE ROOTS:
~ (!) Can now change stance by pressing taunt. Up taunt activates standard plant, side taunt activates prickly plant, and down taunt activates putrid plant.
~~ Putrid:
~ (!) Moves at 88% speed
~ (!) Head attacks have 60% power, bite attacks have 85% power, and other attacks have 70% power
~ (!) Bite attacks deal poison DoT damage
~~ Prickly:
~ (!) Moves at 75% speed
~ (!) Head attacks have 1.4x power, bite attacks have 1.05x power, and other attacks have 1.2x power
~ (!) Attacks have increased startup

Jab [Other]:
~~ Jab 1:
~ [+] Hitbox Size (ID0/1/2&3): 1.8u/1.8u/2.0u → 2.5u/2.5u/3.0u
~~ Jab 2:
~ [+] Hitbox Size (ID0/1/2&3): 2.6u/2.8u/3.0u → 3.2u/3.4u/4.0u
~ (!) Prickly cannot do rapid jab
~~ Jab 3 [Head]:
~~~ Prickly:
~ Startup: F10
~~ Rapid Jab Finisher [Bite]:
~~~ Putrid:
~ Poison DoT: 2%

Forward Tilt [Bite]:
~~ Hit 1:
~~~ Prickly:
~ Startup: F10
~~ Hit 2:
~~~ Putrid:
~ Poison DoT: 3.0%
~~~ Prickly:
~ Startup: F8

Up Tilt [Head]:
~~~ Prickly:
~ Startup: F11

Down Tilt [Other]:
~ [-] Startup: → F8
~~~ Putrid:
~ Knockback compensated
~~~ Prickly:
~ Startup: F11

Dash Attack [Other]:
~ [-] Hitbox Duration: F7-F22 -> F7-F19
~~ Late Hit: 
~ [-] Angle: 85 → 50 
~~~ Prickly:
~ Startup: F11

Forward Smash [Head]:
~ [-] Endlag increased (F63? motion rate removed on endlag)
~~~ Prickly:
~ Startup: F26

Up Smash [Bite]:
~~~ Putrid:
~ Poison DoT: 4.0%
~~~ Prickly:
~ Startup: F16

Down Smash [Other]:
~ [-] Startup: → F14
~~~ Prickly:
~ Startup: F19

Neutral Air [Other]:
~~ Multihits: 
~ [-] Damage: 2.5% → 2% 
~ [-] Angle: 367 → 365
~~~ Prickly:
~ Startup: F8

Forward Air [Other]:
~ [-] Intangibility removed
~ [-] Startup: → F9
~ [+] Interpolation added
~~~ Prickly:
~ Startup: F13

Back Air [Other]:
~ [/] Angle: 361 → 50
~~~ Prickly:
~ Startup: F18
~~~ Putrid:
~ Reworked into a multihit
~~ Multihit (F14 → F25)
~ Size: 7.0u
~ Damage: 1.4%
~ Angle: 366
~ KBG: 100
~ BKB: 15
~~ Launcher (F26 → F30)
~ Size: 9.0u
~ Damage: 4.9%
~ Angle: 50
~ KBG: 108
~ BKB: 25
~ Poison DoT: 3.0%

Up Air [Head]:
~ [-] Intangibility removed
~ [-] Startup: F6 → F9
~~~ Putrid:
~ Knockback compensated
~ Angle: 85 → 60 
~~~ Prickly:
~ Startup: F11

Down Air [Other]:
~ [-] Startup: F9 → F11
~ [/] KBG (ID3) fixed: 90 → 64
~~~ Prickly:
~ Startup: F14

Forward Throw [Head]:
~~~ Prickly:
~ Startup: F17

Down Throw [Bite]:
~~~ Putrid:
~ Poison DoT: 0.5%

Ptooie:
~ [-] Damage (Weak/Strong): 14%/18% → 12%/15%
~~~ Putrid:
~ Damage (Weak/Strong): 10%/12%
~ Poison DoT: 4%
~ Does not bounce
~~~ Prickly:
~ Damage & Size increased by 1.3x
~ Shoot distance reduced

Poison Breath → Adaptive Maw:
~ (!) Now a close-range spit that changes properties depending on the stance
~ Hitbox Size: 9.0u
~~~ Standard:
~ Hitbox Duration: F9-14
~ Damage (Uncharged/Charged): 8%/12%
~ BKB: 60
~ KBG: 66
~ Angle: 30
~~~ Putrid:
~~ Multihits:
~ Hitbox Duration: F8-24
~ Damage (Uncharged/Charged): 1.5%/2.5%
~ BKB: 10
~ KBG: 66
~ Angle: 366
~~ Finisher:
~ Hitbox Duration: F24-27
~ Damage (Uncharged/Charged): 3%/2%
~ BKB: 70
~ KBG: 66
~ Angle: 50
~ Poison DoT (Uncharged/Charged): 3.0%/5.0%
~~~ Prickly:
~ Hitbox Duration: F13-16
~ Damage (Uncharged/Charged): 10%/16%
~ FKB: 30
~ KBG: 100
~ Angle: 290
~ 5.0 shield damage uncharged, ignores shield at full charge

Piranhacopter:
~ [-] Startup: → F10
~ [-] Now remains grounded when used on the ground
~ [-] Vertical speed: 1.7 → 1.0
~ [-] Horizontal speed: 2.4 → 1.4
~ [R] Aerial cancel now transitions straight to special fall
~ [R] Multihits removed
New hitbox data:
Strong hit (F10 → F33)
~ Damage: 10%
~ Angle: 30°/40° (grounded/aerial)
~ BKB: 70/50 (grounded/aerial)
~ KBG: 70
~ Hitlag: 1.0
~ Hitbox size: 3.5/3.5/5.0/5.0 (ID 0/1/2/3)
~ Hitbox X Position: 6.0/6.0/1.5/1.5 (ID 0/1/2/3)
Late hit (F33 → F65)
~ Damage: 7%
~ Angle: 40°/50° (grounded/aerial)
~ BKB: 70/50 (grounded/aerial)
~ KBG: 70
~ Hitlag: 1.0
~ Hitbox size: 3.5/3.5/5.0/5.0 (ID 0/1/2/3)
~ Hitbox X Position: 6.0/6.0/1.5/1.5 (ID 0/1/2/3)

Long-Stem Strike:
~~~ Putrid:
~ Poison DoT (Uncharged/Charged): 2.0%/3.0%

### CODE REVIEW

**Comment by Detronlegacy** [comment]
Wavedashes allow plant to ignore speed reductions caused by stance switching.

**Comment by WoodrowManuel** [comment]
Putrid Ptooie's initial hitbox was moved from the lips to the base of the pot.
https://streamable.com/5cnggo

File: fighters/packun/src/opff.rs [hunk 1]
```diff
@@ -4,22 +4,153 @@ use super::*;
...
+        VarModule::set_float(fighter.object(), vars::common::instance::JUMP_SPEED_MAX_MUL, 0.88);
+
+        // if you are initial dash, slow them down slightly
+        if fighter.is_status(*FIGHTER_STATUS_KIND_DASH) {
+            let motion_vec = Vector3f {
+                x: -0.15 * PostureModule::lr(fighter.boma()) * (1.0 - (MotionModule::frame(fighter.boma()) / MotionModule::end_frame(fighter.boma()))),
+                y: 0.0, 
+                z: 0.0
+            };
+            KineticModule::add_speed_outside(fighter.boma(), *KINETIC_OUTSIDE_ENERGY_TYPE_WIND_NO_ADDITION, &motion_vec);
```
Comment (jobrien97) [inline fighters/packun/src/opff.rs]: Please don't use wind, dash should already be covered in the control energy status speed mul, no?

File: fighters/packun/src/opff.rs [hunk 2]
```diff
@@ -4,22 +4,153 @@ use super::*;
...
+        *FIGHTER_STATUS_KIND_DEAD]) {
+            VarModule::set_int(fighter.object(), vars::packun::instance::CURRENT_STANCE, 0);
+    }
+}
+
+/// applies the given multiplier to various speed stats of the given fighter. 
+/// This should only be called once per status, or you will get some multiplicative effects
+unsafe fn apply_status_speed_mul(fighter: &mut smash::lua2cpp::L2CFighterCommon, mul: f32) {
+    // set the X motion speed multiplier (where movement is baked into an anim)
+    lua_bind::FighterKineticEnergyMotion::set_speed_mul(fighter.get_motion_energy(), mul);
```
Comment (jobrien97) [inline fighters/packun/src/opff.rs]: Please do
```
fighter.clear_lua_stack();
lua_args!(fighter, FIGHTER_KINETIC_ENERGY_ID_MOTION);
let og_speed_mul = app::sv_kinetic_energy::get_speed_mul(fighter.lua_state_agent);
lua_bind::FighterKineticEnergyMotion::set_speed_mul(fighter.get_motion_energy(), og_speed_mul * mul);
```
instead. This allows dash attack to be naturally affected by your status speed mul.

File: fighters/packun/src/acmd/ground.rs [hunk 3]
```diff
@@ -65,45 +75,78 @@ unsafe fn packun_attack_12_game(fighter: &mut L2CAgentBase) {
...
+
 
 #[acmd_script( agent = "packun", script = "game_attackdash" , category = ACMD_GAME , low_priority)]
 unsafe fn packun_attack_dash_game(fighter: &mut L2CAgentBase) {
     let lua_state = fighter.lua_state_agent;
     let boma = fighter.boma();
+    let stance = StanceInfo::from(VarModule::get_int(boma.object(), vars::packun::instance::CURRENT_STANCE));
     // base 0.93
-    sv_kinetic_energy!(set_speed_mul, fighter, FIGHTER_KINETIC_ENERGY_ID_MOTION, 0.8);
+    sv_kinetic_energy!(set_speed_mul, fighter, FIGHTER_KINETIC_ENERGY_ID_MOTION, stance.da_speed);
```
Comment (jobrien97) [inline fighters/packun/src/acmd/ground.rs]: Please revert this in favor of what I commented in opff.rs

**Review by jobrien97 — DISMISSED** [review]

**Review by SkewedAskew — APPROVED** [review]

**Pull request merged**